File: backend/cara/schemas/diet.py
"""Pydantic schemas for the CARA Nutrizione (diet) module."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

# ─── Catalog / rules ───────────────────────────────────────────────


class FoodItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    status: str
    protein_category: str | None
    food_group: str | None
    default_portion_g: int | None
    portion_primo_g: int | None
    portion_secondo_g: int | None
    kcal_per_100g: int | None
    season_months: list[int] | None
    notes: str | None


class DietRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan_id: int
    category: str
    target_min: float | None
    target_max: float | None
    period: str | None
    portion_primo_g: int | None
    portion_secondo_g: int | None
    portion_note: str | None
    kcal_estimate: int | None


class DietRuleUpdate(BaseModel):
    """Editable fields for a rule (admin/UI). All optional → patch."""

    target_min: float | None = None
    target_max: float | None = None
    period: str | None = None
    portion_primo_g: int | None = None
    portion_secondo_g: int | None = None
    portion_note: str | None = None
    kcal_estimate: int | None = None


class RecipeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    ingredients: list
    steps: str | None
    source: str | None


# ─── Meal logging ──────────────────────────────────────────────────


class ParsedItem(BaseModel):
    food: str
    portion_g: int | None = None
    protein_category: str | None = None
    food_group: str | None = None
    status: str | None = None


class MealLogIn(BaseModel):
    free_text: str = Field(..., min_length=1, max_length=1000)
    meal_type: str | None = Field(
        default=None, description="colazione|spuntino|pranzo|cena; inferred if omitted"
    )
    logged_at: datetime | None = None


class MealLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    logged_at: datetime
    meal_type: str
    free_text: str | None
    parsed_items: list
    est_kcal: int | None
    protein_category: str | None
    context_flags: dict


class MealLogResult(BaseModel):
    """POST /diet/log response — the saved row + parser warnings."""

    log: MealLogOut
    warnings: list[str] = Field(default_factory=list)
    carb_present: bool = False
    vegetable_present: bool = False
    fruit_present: bool = False


# ─── Today view ────────────────────────────────────────────────────


class TodayMealSlot(BaseModel):
    meal_type: str
    done: bool
    logs: list[MealLogOut] = Field(default_factory=list)


class TodayOut(BaseModel):
    day: date
    slots: list[TodayMealSlot]
    missing: list[str]
    water_ml: int
    water_target_min: int
    water_target_max: int
    coffee_count: int
    coffee_max: int
    fruit_servings: int
    fruit_target_min: int


# ─── Week / frequencies ────────────────────────────────────────────


class CategoryProgress(BaseModel):
    category: str
    color: str
    consumed: float
    target_min: float | None
    target_max: float | None
    state: str  # ok | under | over | warn


class WeekOut(BaseModel):
    week_start: date
    week_end: date
    categories: list[CategoryProgress]
    adherence_score: float
    avg_kcal_per_day: int | None
    avg_water_ml: int | None
    avg_coffee: float | None
    notes: list[str] = Field(default_factory=list)


# ─── Suggestions ───────────────────────────────────────────────────


class DishSuggestion(BaseModel):
    title: str
    covers_category: str | None
    detail: str


class SuggestOut(BaseModel):
    meal_type: str
    suggestions: list[DishSuggestion]
    under_target: list[str]
    warnings: list[str]
    context_reminders: list[str]
    speak_text: str


# ─── Intake counters ───────────────────────────────────────────────


class WaterIn(BaseModel):
    ml: int = Field(default=250, ge=0, le=3000)


class CoffeeIn(BaseModel):
    count: int = Field(default=1, ge=-5, le=10)


class IntakeOut(BaseModel):
    day: date
    water_ml: int
    coffee_count: int


# ─── Weekly report ─────────────────────────────────────────────────


class WeeklyReportOut(BaseModel):
    week_start: date
    week_end: date
    pdf_url: str | None
    telegram_text: str
    sent_to_telegram: bool


# ─── Energy: profile / exercise / stats ────────────────────────────


class DietProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sex: str | None
    birth_date: date | None
    height_cm: float | None
    weight_kg: float | None
    activity_level: str
    goal: str
    goal_rate_kcal: int | None
    # Derived (computed, not stored):
    age: int | None = None
    bmr: int | None = None
    tdee: int | None = None
    daily_target: int | None = None
    complete: bool = False


class DietProfileUpdate(BaseModel):
    sex: str | None = Field(default=None, pattern="^[MFmf]$")
    birth_date: date | None = None
    height_cm: float | None = Field(default=None, ge=80, le=250)
    weight_kg: float | None = Field(default=None, ge=25, le=350)
    activity_level: str | None = Field(
        default=None, pattern="^(sedentary|light|moderate|very|extra)$"
    )
    goal: str | None = Field(default=None, pattern="^(maintain|lose|gain)$")
    goal_rate_kcal: int | None = Field(default=None, ge=-1500, le=1500)


class MetActivityOut(BaseModel):
    slug: str
    label: str
    met: float
    intensity: str


# ─── Barcode (Open Food Facts) ─────────────────────────────────────


class BarcodeProductOut(BaseModel):
    """A product resolved from a scanned barcode (Open Food Facts)."""

    barcode: str
    name: str
    brand: str | None = None
    quantity: str | None = None       # as printed on the pack, e.g. "500 g"
    kcal_per_100g: int | None = None
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None
    fiber_g: float | None = None
    nutriscore: str | None = None     # a..e
    image_url: str | None = None
    # True if it came from our local catalog cache (instant/offline),
    # False if freshly fetched from Open Food Facts.
    from_cache: bool = False


class BarcodeLogIn(BaseModel):
    barcode: str = Field(..., min_length=6, max_length=14)
    portion_g: int = Field(..., ge=1, le=2000)
    meal_type: str | None = Field(
        default=None, description="colazione|spuntino|pranzo|cena; inferred if omitted"
    )
    logged_at: datetime | None = None


# ─── "Cosa cucino?" — piatti dalla spesa ───────────────────────────


class DishIngredientOut(BaseModel):
    name: str
    portion_g: int | None = None
    kcal: int | None = None
    have: bool = False
    status: str | None = None


class DishProposalOut(BaseModel):
    slug: str
    title: str
    covers_category: str | None = None
    kcal_estimate: int | None = None
    ingredients: list[DishIngredientOut] = Field(default_factory=list)
    missing: list[str] = Field(default_factory=list)
    note: str | None = None


class RecipeIngredient(BaseModel):
    item: str
    qty: str | None = None


class RecipeDetailOut(BaseModel):
    name: str
    ingredients: list[RecipeIngredient] = Field(default_factory=list)
    steps: str | None = None
    source: str | None = None


class DishShoppingIn(BaseModel):
    names: list[str] = Field(..., min_length=1, max_length=20)


class ExerciseLogIn(BaseModel):
    activity: str = Field(..., min_length=1, max_length=80)
    duration_min: int = Field(..., ge=1, le=600)
    # Provide either a catalog slug (MET looked up) or an explicit met.
    slug: str | None = None
    met: float | None = Field(default=None, ge=0.5, le=20)
    logged_at: datetime | None = None
    notes: str | None = None


class ExerciseLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    logged_at: datetime
    activity: str
    met: float
    duration_min: int
    kcal_burned: int
    notes: str | None


class EnergyBucket(BaseModel):
    label: str
    start: str
    consumed: int
    burned: int
    target: int | None


class EnergyStatsOut(BaseModel):
    period: str
    period_start: date
    period_end: date
    profile_complete: bool
    bmr: int | None
    tdee: int | None
    daily_target: int | None
    days_elapsed: int
    consumed: int
    burned: int
    net: int
    target_total: int | None
    remaining: int | None
    avg_consumed_per_day: int
    avg_burned_per_day: int
    breakdown: list[EnergyBucket]

File: frontend/src/api/diet.ts
// API client for the CARA Nutrizione (diet) module.
//
// Router: /api/v1/diet. Auth via the shared authFetch. The plan reasons
// by WEEKLY FREQUENCIES of protein categories + raw portions; calories
// are a secondary metric. Disclaimer: reminders/suggestions only, not a
// medical device.

import { authFetch } from './auth';

const API = '/api/v1/diet';

export type FoodStatus = 'consigliato' | 'da_moderare' | 'sconsigliato';
export type ProteinCategory = 'legumi' | 'pesce' | 'carne' | 'uova' | 'formaggio';
export type MealType = 'colazione' | 'spuntino' | 'pranzo' | 'cena';
export type CategoryState = 'ok' | 'under' | 'over' | 'warn';

// Plan colors (from the PDF, page 12).
export const CATEGORY_COLOR: Record<ProteinCategory, string> = {
  legumi: '#22c55e', // verde
  pesce: '#38bdf8', // azzurro
  carne: '#d946ef', // magenta
  uova: '#3b82f6', // blu
  formaggio: '#eab308', // giallo
};

export const STATUS_LABEL: Record<FoodStatus, string> = {
  consigliato: 'Consigliato',
  da_moderare: 'Da moderare',
  sconsigliato: 'Sconsigliato',
};

export interface FoodItem {
  id: number;
  name: string;
  status: FoodStatus;
  protein_category: ProteinCategory | null;
  default_portion_g: number | null;
  portion_primo_g: number | null;
  portion_secondo_g: number | null;
  kcal_per_100g: number | null;
  season_months: number[] | null;
  notes: string | null;
}

export interface ParsedItem {
  food: string;
  portion_g: number | null;
  protein_category: ProteinCategory | null;
  status: FoodStatus | null;
  known?: boolean;
}

export interface MealLog {
  id: number;
  user_id: number;
  logged_at: string;
  meal_type: MealType;
  free_text: string | null;
  parsed_items: ParsedItem[];
  est_kcal: number | null;
  protein_category: ProteinCategory | null;
  context_flags: Record<string, boolean>;
}

export interface MealLogResult {
  log: MealLog;
  warnings: string[];
  carb_present: boolean;
  vegetable_present: boolean;
  fruit_present: boolean;
}

export interface TodayMealSlot {
  meal_type: MealType;
  done: boolean;
  logs: MealLog[];
}

export interface Today {
  day: string;
  slots: TodayMealSlot[];
  missing: MealType[];
  water_ml: number;
  water_target_min: number;
  water_target_max: number;
  coffee_count: number;
  coffee_max: number;
  fruit_servings: number;
  fruit_target_min: number;
}

export interface CategoryProgress {
  category: ProteinCategory;
  color: string;
  consumed: number;
  target_min: number | null;
  target_max: number | null;
  state: CategoryState;
}

export interface Week {
  week_start: string;
  week_end: string;
  categories: CategoryProgress[];
  adherence_score: number;
  avg_kcal_per_day: number | null;
  avg_water_ml: number | null;
  avg_coffee: number | null;
  notes: string[];
}

export interface DishSuggestion {
  title: string;
  covers_category: ProteinCategory | null;
  detail: string;
}

export interface Suggest {
  meal_type: MealType;
  suggestions: DishSuggestion[];
  under_target: ProteinCategory[];
  warnings: string[];
  context_reminders: string[];
  speak_text: string;
}

export interface DietRule {
  id: number;
  plan_id: number;
  category: string;
  target_min: number | null;
  target_max: number | null;
  period: string | null;
  portion_primo_g: number | null;
  portion_secondo_g: number | null;
  portion_note: string | null;
  kcal_estimate: number | null;
}

export interface Recipe {
  id: number;
  name: string;
  ingredients: { item: string; qty: string }[];
  steps: string | null;
  source: string | null;
}

export interface Intake {
  day: string;
  water_ml: number;
  coffee_count: number;
}

export interface WeeklyReport {
  week_start: string;
  week_end: string;
  pdf_url: string | null;
  telegram_text: string;
  sent_to_telegram: boolean;
}

async function json<T>(r: Response): Promise<T> {
  if (!r.ok) {
    const detail = await r
      .json()
      .then((j) => j.detail)
      .catch(() => `HTTP ${r.status}`);
    throw new Error(typeof detail === 'string' ? detail : `HTTP ${r.status}`);
  }
  return r.json();
}

export async function logMeal(payload: {
  free_text: string;
  meal_type?: MealType;
}): Promise<MealLogResult> {
  return json(
    await authFetch(`${API}/log`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

// ─── Barcode (Open Food Facts) ─────────────────────────────────────

export interface BarcodeProduct {
  barcode: string;
  name: string;
  brand: string | null;
  quantity: string | null;
  kcal_per_100g: number | null;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
  fiber_g: number | null;
  nutriscore: string | null; // a..e
  image_url: string | null;
  from_cache: boolean;
}

/** Resolve a scanned barcode. Throws "Prodotto non trovato" on 404. */
export async function lookupBarcode(code: string): Promise<BarcodeProduct> {
  return json(await authFetch(`${API}/barcode/${encodeURIComponent(code)}`));
}

/** Log a scanned product as a meal (kcal computed from portion). */
export async function logBarcode(payload: {
  barcode: string;
  portion_g: number;
  meal_type?: MealType;
}): Promise<MealLog> {
  return json(
    await authFetch(`${API}/barcode/log`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

// ─── "Cosa cucino?" — piatti dalla spesa ───────────────────────────

export interface DishIngredient {
  name: string;
  portion_g: number | null;
  kcal: number | null;
  have: boolean;
  status: FoodStatus | null;
}

export interface DishProposal {
  slug: string;
  title: string;
  covers_category: ProteinCategory | null;
  kcal_estimate: number | null;
  ingredients: DishIngredient[];
  missing: string[];
  note: string | null;
}

export interface RecipeDetail {
  name: string;
  ingredients: { item: string; qty: string | null }[];
  steps: string | null;
  source: string | null;
}

export async function getDishes(meal: MealType = 'cena'): Promise<DishProposal[]> {
  return json(await authFetch(`${API}/dishes?meal=${meal}`));
}

export async function getDishRecipe(title: string): Promise<RecipeDetail> {
  return json(await authFetch(`${API}/dishes/recipe?title=${encodeURIComponent(title)}`));
}

export async function addDishMissingToShopping(names: string[]): Promise<{ added: number }> {
  return json(
    await authFetch(`${API}/dishes/shopping`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ names }),
    }),
  );
}

export async function getToday(): Promise<Today> {
  return json(await authFetch(`${API}/today`));
}

export interface HistoryDay {
  day: string;
  meals: MealLog[];
  total_kcal: number | null;
}

export interface History {
  days: HistoryDay[];
}

/** Pasti registrati negli ultimi `days` giorni, raggruppati per giorno. */
export async function getHistory(days = 14): Promise<History> {
  return json(await authFetch(`${API}/history?days=${days}`));
}

/** Modifica un pasto: sposta lo slot e/o riscrivi il testo. */
export async function updateMeal(
  id: number,
  patch: { meal_type?: MealType; free_text?: string },
): Promise<MealLogResult> {
  return json(
    await authFetch(`${API}/meal/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(patch),
    }),
  );
}

/** Elimina un pasto registrato. */
export async function deleteMeal(id: number): Promise<void> {
  const r = await authFetch(`${API}/meal/${id}`, { method: 'DELETE' });
  if (!r.ok && r.status !== 204) throw new Error(`HTTP ${r.status}`);
}

export async function getWeek(anchor?: string): Promise<Week> {
  const q = anchor ? `?anchor=${anchor}` : '';
  return json(await authFetch(`${API}/week${q}`));
}

export async function getSuggest(meal: MealType = 'cena'): Promise<Suggest> {
  return json(await authFetch(`${API}/suggest?meal=${meal}`));
}

export async function getFoods(opts?: {
  status?: FoodStatus;
  category?: ProteinCategory;
  q?: string;
}): Promise<FoodItem[]> {
  const p = new URLSearchParams();
  if (opts?.status) p.set('status', opts.status);
  if (opts?.category) p.set('category', opts.category);
  if (opts?.q) p.set('q', opts.q);
  return json(await authFetch(`${API}/foods?${p.toString()}`));
}

export async function getRecipes(): Promise<Recipe[]> {
  return json(await authFetch(`${API}/recipes`));
}

export async function getRules(): Promise<DietRule[]> {
  return json(await authFetch(`${API}/rules`));
}

export async function updateRule(id: number, patch: Partial<DietRule>): Promise<DietRule> {
  return json(
    await authFetch(`${API}/rules/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(patch),
    }),
  );
}

export async function addWater(ml = 250): Promise<Intake> {
  return json(
    await authFetch(`${API}/water`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ml }),
    }),
  );
}

export async function addCoffee(count = 1): Promise<Intake> {
  return json(
    await authFetch(`${API}/coffee`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ count }),
    }),
  );
}

export async function generateWeeklyReport(opts?: {
  shareTelegram?: boolean;
  anchor?: string;
}): Promise<WeeklyReport> {
  const p = new URLSearchParams();
  if (opts?.shareTelegram) p.set('share_telegram', 'true');
  if (opts?.anchor) p.set('anchor', opts.anchor);
  return json(await authFetch(`${API}/report/weekly?${p.toString()}`, { method: 'POST' }));
}

// ─── Energy: profile / exercise / stats ────────────────────────────

export type ActivityLevel = 'sedentary' | 'light' | 'moderate' | 'very' | 'extra';
export type Goal = 'maintain' | 'lose' | 'gain';
export type EnergyPeriod = 'day' | 'week' | 'month' | 'year';

export interface DietProfile {
  sex: 'M' | 'F' | null;
  birth_date: string | null;
  height_cm: number | null;
  weight_kg: number | null;
  activity_level: ActivityLevel;
  goal: Goal;
  goal_rate_kcal: number | null;
  age: number | null;
  bmr: number | null;
  tdee: number | null;
  daily_target: number | null;
  complete: boolean;
}

export interface MetActivity {
  slug: string;
  label: string;
  met: number;
  intensity: string;
}

export interface ExerciseLog {
  id: number;
  user_id: number;
  logged_at: string;
  activity: string;
  met: number;
  duration_min: number;
  kcal_burned: number;
  notes: string | null;
}

export interface EnergyBucket {
  label: string;
  start: string;
  consumed: number;
  burned: number;
  target: number | null;
}

export interface EnergyStats {
  period: EnergyPeriod;
  period_start: string;
  period_end: string;
  profile_complete: boolean;
  bmr: number | null;
  tdee: number | null;
  daily_target: number | null;
  days_elapsed: number;
  consumed: number;
  burned: number;
  net: number;
  target_total: number | null;
  remaining: number | null;
  avg_consumed_per_day: number;
  avg_burned_per_day: number;
  breakdown: EnergyBucket[];
}

export const ACTIVITY_LABEL: Record<ActivityLevel, string> = {
  sedentary: 'Sedentario',
  light: 'Leggermente attivo',
  moderate: 'Moderatamente attivo',
  very: 'Molto attivo',
  extra: 'Estremamente attivo',
};

export const GOAL_LABEL: Record<Goal, string> = {
  maintain: 'Mantenere il peso',
  lose: 'Perdere peso',
  gain: 'Aumentare di peso',
};

export const PERIOD_LABEL: Record<EnergyPeriod, string> = {
  day: 'Giorno',
  week: 'Settimana',
  month: 'Mese',
  year: 'Anno',
};

export async function getProfile(): Promise<DietProfile> {
  return json(await authFetch(`${API}/profile`));
}

export async function putProfile(patch: Partial<DietProfile>): Promise<DietProfile> {
  return json(
    await authFetch(`${API}/profile`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(patch),
    }),
  );
}

export async function getEnergy(period: EnergyPeriod = 'day'): Promise<EnergyStats> {
  return json(await authFetch(`${API}/energy?period=${period}`));
}

export async function getExerciseCatalog(): Promise<MetActivity[]> {
  return json(await authFetch(`${API}/exercise/catalog`));
}

export async function listExercise(day?: string): Promise<ExerciseLog[]> {
  const q = day ? `?day=${day}` : '';
  return json(await authFetch(`${API}/exercise${q}`));
}

export async function addExercise(payload: {
  activity: string;
  slug?: string;
  met?: number;
  duration_min: number;
  notes?: string;
}): Promise<ExerciseLog> {
  return json(
    await authFetch(`${API}/exercise`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

export async function deleteExercise(id: number): Promise<void> {
  const r = await authFetch(`${API}/exercise/${id}`, { method: 'DELETE' });
  if (!r.ok && r.status !== 204) throw new Error(`HTTP ${r.status}`);
}

export const DISCLAIMER =
  'Promemoria e suggerimenti, non un dispositivo medico. La fonte è il piano ' +
  'della Dott.ssa Elena Poletti; per dubbi rivolgersi a lei.';

export const MEAL_LABEL: Record<MealType, string> = {
  colazione: 'Colazione',
  spuntino: 'Spuntino',
  pranzo: 'Pranzo',
  cena: 'Cena',
};
